jdaviz/configs/cubeviz/plugins/slice/slice.py

```diff
@@ -0,0 +1,132 @@
+        #  in _all_ viewers should be uniform.
+        self.session.hub.subscribe(self, AddDataMessage,
+                                   handler=self._on_data_added)
+
+    @observe("linked")
+    def _on_linked_changed(self, event):
+        for viewer in self._watched_viewers:
+
+            if not event['new']:
+                viewer.state.remove_callback('slices',
+                                             self._viewer_slices_changed)
+            else:
+                viewer.state.add_callback('slices',
+                                          self._viewer_slices_changed)
+
+    def _on_data_added(self, msg):
+        if len(msg.data.shape) == 3 and \
+                isinstance(msg.viewer, BqplotImageView):
+            self.max_value = msg.data.shape[0] - 1
+
+            if msg.viewer not in self._watched_viewers:
+                self._watched_viewers.append(msg.viewer)
+
+                msg.viewer.state.add_callback('slices',
+                                              self._viewer_slices_changed)
+
+        elif isinstance(msg.viewer, BqplotProfileView):
+            if msg.viewer not in self._indicator_viewers:
+                self._indicator_viewers.append(msg.viewer)
+                # cache wavelengths so that wavelength <> slice conversion can be done efficiently
+                x_all = msg.viewer.data()[0].spectral_axis
+                self._x_all = x_all.value
+                self.wavelength_unit = str(x_all.unit)
+                # but if the units (or data) change, we need to update internally
+                msg.viewer.state.add_callback("reference_data",
+                                              lambda reference_data: self._update_data(reference_data.get_object().spectral_axis)) # noqa
+
+                if self.wavelength == -1:
+                    # initialize at middle of cube
+                    self.slider = int(len(x_all)/2)
+
+    def _update_data(self, x_all):
+        if hasattr(x_all, 'unit'):
+            self.wavelength_unit = str(x_all.unit)
+            x_all = x_all.value
+
+        self._x_all = x_all
+        # force wavelength to update from the current slider value
+        self._on_slider_updated({'new': self.slider})
+
+    def _viewer_slices_changed(self, value):
+        # the slices attribute on the viewer state was changed,
+        # so we'll update the slider to match which will trigger
+        # the slider observer (_on_slider_updated) and sync across
+        # any other applicable viewers
+        if len(value):
+            self.slider = float(value[0])
+
+    def _on_select_slice_message(self, msg):
+        # NOTE: by setting the slice index, the observer (_on_slider_updated)
+        # will sync across all viewers and update the wavelength
+        self.slider = msg.slice
+
+    def vue_change_wavelength(self, event):
+        # convert to float after stripping any invalid characters
+        value = float(re.sub(r'[^-+eE\d.]', '', event))
```

Comment:
Well, a number input doesn't work because (I think) we want to support sci notation, which leaves us either with validation, error, or stripping.  My preference is always to "bother" the user as little as possible since there will be immediate feedback of any changes, but we probably should decide on a policy app-wide for situations like this and be consistent (it has also come up in a few other recent PRs).